File: examples/demo-server/scripts/session/regen.lua
-- session/regen.lua — Stat regeneration and meditation system.
--
-- Provides periodic HP / Mana / Stamina regeneration for the player session.
-- Meditation is an active skill (skill_id=46) that boosts mana regen and
-- is interrupted by damage, casting, or other actions.
--
-- Depends on: constants.lua (REGEN table)

-- ══════════════════════════════════════════════════════════════════════════
-- Skill IDs
-- ══════════════════════════════════════════════════════════════════════════

SKILL_MEDITATION = 46

-- ══════════════════════════════════════════════════════════════════════════
-- Meditation state
-- ══════════════════════════════════════════════════════════════════════════

is_meditating = false

--- Start meditation. Returns a system message.
function start_meditation()
    is_meditating = true
    return "You enter a meditative trance."
end

--- Stop meditation if active. Returns a message or nil.
function stop_meditation()
    if is_meditating then
        is_meditating = false
        return "You stop meditating."
    end
    return nil
end

--- Toggle meditation (UseSkill handler).
function toggle_meditation()
    if is_meditating then
        local msg = stop_meditation()
        if msg then
            session:send_system_message(msg)
        end
    else
        local msg = start_meditation()
        session:send_system_message(msg)
    end
end

--- Interrupt meditation due to action/damage.
--- Sends a system message if meditation was active.
function interrupt_meditation()
    local msg = stop_meditation()
    if msg then
        session:send_system_message(msg)
    end
end

-- ══════════════════════════════════════════════════════════════════════════
-- Regen tick
-- ══════════════════════════════════════════════════════════════════════════

--- Perform one regen tick: restore HP, mana (with meditation bonus), stamina.
function regen_tick()
    local my_serial = session:player_serial()

    -- HP regen
    engine:heal_entity(my_serial, REGEN.HP_PER_TICK)

    -- Mana regen (boosted by meditation)
    local mana_amount = REGEN.MANA_PER_TICK
    if is_meditating then
        mana_amount = mana_amount + REGEN.MEDITATION_BONUS
    end
    engine:modify_mana(my_serial, mana_amount)

    -- Stamina regen
    engine:modify_stamina(my_serial, REGEN.STAM_PER_TICK)
end

File: examples/demo-server/regen.rs
//! Stat regeneration and meditation system.
//!
//! Provides periodic HP / Mana / Stamina regeneration for player sessions.
//! Meditation is an active skill that boosts mana regen while active and
//! is interrupted by damage or actions.
//!
//! ## Usage
//!
//! The session loop calls [`RegenState::tick`] on a periodic timer
//! (every [`regen::TICK_INTERVAL_MS`]).
//! Meditation is toggled via [`RegenState::start_meditation`] /
//! [`RegenState::stop_meditation`].

use crate::constants::regen;
use crate::DemoWorkerTx;

// ── MeditationState ───────────────────────────────────────────────────────

/// Whether the player is actively meditating.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MeditationState {
    /// Not meditating.
    Inactive,
    /// Actively meditating (bonus mana regen).
    Active,
}

// ── RegenState ────────────────────────────────────────────────────────────

/// Per-session regeneration state.
pub struct RegenState {
    pub meditation: MeditationState,
}

impl RegenState {
    pub fn new() -> Self {
        Self {
            meditation: MeditationState::Inactive,
        }
    }

    /// Returns `true` if the player is actively meditating.
    pub fn is_meditating(&self) -> bool {
        self.meditation == MeditationState::Active
    }

    /// Begin meditation. Returns a system message to send to the client.
    pub fn start_meditation(&mut self) -> &'static str {
        self.meditation = MeditationState::Active;
        "You enter a meditative trance."
    }

    /// End meditation (e.g. due to damage, action, or manual toggle).
    /// Returns `Some(message)` if meditation was actually active.
    pub fn stop_meditation(&mut self) -> Option<&'static str> {
        if self.meditation == MeditationState::Active {
            self.meditation = MeditationState::Inactive;
            Some("You stop meditating.")
        } else {
            None
        }
    }

    /// Perform one regen tick. Sends engine commands to heal/restore stats.
    ///
    /// `serial` is the player entity serial, `world` is the map ID.
    /// Returns packets to send to the session (currently none, but
    /// reserved for future "mana full" notifications).
    pub async fn tick(
        &self,
        serial: u32,
        world: u8,
        worker_tx: &DemoWorkerTx,
    ) {
        let engine = crate::game_util::engine_for(worker_tx, world);

        // HP regen
        let _ = engine.heal(serial, regen::HP_PER_TICK).await;

        // Mana regen (boosted by meditation)
        let mana_amount = if self.is_meditating() {
            regen::MANA_PER_TICK + regen::MANA_MEDITATION_BONUS
        } else {
            regen::MANA_PER_TICK
        };
        let _ = engine.modify_mana(serial, mana_amount as i32).await;

        // Stamina regen
        let _ = engine.modify_stamina(serial, regen::STAM_PER_TICK as i32).await;
    }
}
